### PR TITLE
Feat: parse translated challenges

### LINKED ISSUE #37599: Need to have consistent comments (and translations of those comments) throughout curriculum files
A PR (#37111 ) was recently merged which adds the following comment to the Solutions section of all the back-end challenges. I just saw a new PR (#37592) added the untranslated comment to a non-English challenge. This got me to thinking about more a general question: shouldn't all comments regardless where they appear in the challenge (i.e. Solutions section, example code in description or instructions, or challenge seed code) be translated to the applicable language?

By Solutions section, I mean the comment like `// solution required` or the new multi-line comment like the one below:
````
/**
Backend challenges don't need solutions, 
because they would need to be tested against a full working project. 
Please check our contributing guidelines to learn more.
*/
```
````
For at least the Solutions section, we need to inquire with the community to find out the "correct" translation for that particular comment and then see if we can limit the comments in the challenge seed section to only a few variances and then have the community tell us the "correct" translation. By "correct", I mean not just using Google to create a machine translation that might have to be tweaked several times in various PRs for each language.

For the challenge seed code, we would need to try and limit the number of unique comments.  In just the challenge seed code, I have found the following comment variations the curriculum in JavaScript code:

````
//en.wikipedia.org/wiki/Larry_Page" target="_blank" class="links">Larry Page</a><br><br>

//en.wikipedia.org/wiki/Sergey_Brin" target="_blank" class="links">Sergey Brin</a>

//en.wikipedia.org/wiki/Larry_Page" target="_blank" class="links">Larry Page</a>

//freecatphotoapp.com/" target="_blank">CatPhotoApp</a>

/* add your code below this line */

/* add your code above this line */

//fonts.googleapis.com/css?family=Lobster" rel="stylesheet" type="text/css">

//bit.ly/fcc-relaxing-cat" alt="A cute orange cat lying on its back."></a>

/* change code below */

/* change code above */

/* add code below */

/* add code above */

//bit.ly/fcc-relaxing-cat" alt="A cute orange cat lying on its back.">

//freecatphotoapp.com" target="_blank">cat photos</a>

//freecatphotoapp.com" target="_blank">cat photos</a>.</p>

//pbs.twimg.com/profile_images/378800000147359764/54dc9a5c34e912f34db8662d53d16a39_400x400.png" alt="Quincy Larson&apos;s profile picture" class="profile-thumbnail">

/* change the code below this line */

/* change the code above this line */

/* enter your code below this line */

/* enter your code above this line */

/* change code below this line */

/* change code above this line */

/* modify the code below this line */

/* modify the code above this line */

/* To avoid rounding errors the largest allowed diff is set at 0.01px */

/* Add media query below */

//s3.amazonaws.com/freecodecamp/FCCStickerPack.jpg" alt="freeCodeCamp stickers set">

//s3.amazonaws.com/freecodecamp/FCCStickers-CamperBot200x200.jpg" alt="freeCodeCamp sticker that says &apos;Because CamperBot Cares&apos;">

// What is the new fad diet for ghost developers? The Boolean.

// Break it up.

// &quot;Never give up and good luck will find you.&quot;

// -- Falcor

// Don&apos;t show a false ID to this bouncer.

// repeat after me

// You can do this!

// It&apos;s alive. It&apos;s alive!

// Clear out that junk in your trunk

// Find my place in this sorted array.

// change code below this line

//change code above this line

// do not change code above this line

// change code below this line to test different cases:

// change code above this line

// do not change code below this line

// change code here to test different cases:

// change this line

// Example

// equals 50

// Setup

// Only change code below this line.

// Only change code below this line

// Change this line

// Only change code below this line;

// Change this Line

// Only change code above this line

// Change this value to test

// Only modify code below this line

// Add/remove calls to test your function.

// Note: Only the last will display

// Declare myName below this line

// Fix this line

// Only change code above this line.

// Change these values to test your function

// Declare your variable here

// Assign 5 to oopsGlobal Here

// Change these values to test

// you shouldn&apos;t need to edit this line

// Run and check the console

// myVar is not defined outside of myLocalScope

// Now remove the console log line to pass the test

// removedFromOurArray now equals 3, and ourArray now equals [1,2]

// ourArray now equals [&quot;Stimpson&quot;, &quot;J&quot;, &quot;cat&quot;, [&quot;happy&quot;, &quot;joy&quot;]]

// removedFromOurArray now equals &quot;Stimpson&quot; and ourArray now equals [&quot;J&quot;, [&quot;cat&quot;]].

// ourArray now equals [&quot;J&quot;, &quot;cat&quot;]

// ourArray now equals [&quot;Happy&quot;, &quot;J&quot;, &quot;cat&quot;]

// Add record here

// ourArray now equals [18,45,99].

// Modify values below to test your code

// Outputs 5

//Setup

//www.example.com\&quot;" target="\&quot;_blank\&quot;">Link</a>&quot;;

// Alter values below to test your code

// Change values below to test your code

// Fix this code

// Your code here

// Test Setup

// Display Code

// Modify this line to test

// Your Code Here

// Test your code by modifying these values

// Fix Me

// Declarations

// Assignments

// Initialize these three variables

// Do not change code below this line

//Only change code below this line

// Only change this line;

// Fix the line below

// Do not alter code below this line

// Open your browser console.

// Use console.clear() on the next line to clear the browser console.

// Use console.log() to print the output variable.

// Check the two consoles to see the difference. The freeCodeCamp console should have printed the variable twice, once for each test of this challenge. The browser console should only print the variable once because you cleared it first.

// Creates a 2-D array with m rows and n columns of zeroes

// Adds the m-th row into newArray

// Pushes n zeroes into the current row to create the columns

// Pushes the current row, which now has n zeroes in it, to the array

// Add your code below this line

// responseFromServer represents a response from a server

/**
 * makeList(result.failure) should return:
 * [ `<li class="text-warning">no-var</li>`,
 *   `<li class="text-warning">var-on-top</li>`,
 *   `<li class="text-warning">linebreak</li>` ]
 **/

// responseFromServer is set to true to represent a successful response from a server

// responseFromServer is set to false to represent an unsuccessful response from a server

// add code above this line

// s = [2, 5, 7]; &lt;- this is invalid

// returns 7

// returns 6

/* Alter code below this line */

/* Alter code above this line */

// =&gt; should be &apos;carrot&apos;

// should be 6

// should be 8

// should be 64

// should be 77

// should be not defined

// should be 80

// use function argument destructuring

// should be object

// should be 28.015

// change this

// should be [3,4,5,6,7,8,9,10]

// should be [1,2,3,4,5,6,7,8,9,10];

// setting in Fahrenheit scale

// 24.44 in C

// 26 in C

// 6

// test your code

// returns a proper object

// Add your code above this line

// the global variable

// Should be &quot;winter-is-coming&quot;

// Should equal 5

// Should print 4

// the global Array

/**
 * A long process to prepare tea.
 * @return {string} A cup of tea.
 **/

/**
 * Get given number of cups of tea.
 * @param {number} numOfCups Number of required cups of tea.
 * @return {Array<string>} Given amount of tea cups.
 **/

// :(

/* This function should add a book to the list and return the list */

// New parameters should come before bookName

/* This function should remove a book from the list and return the list */

// New parameters should come before the bookName one

/**
 * A long process to prepare green tea.
 * @return {string} A cup of green tea.
 **/

/**
 * A long process to prepare black tea.
 * @return {string} A cup of black tea.
 **/

/**
 * Get given number of cups of tea.
 * @param {function():string} prepareTea The type of tea preparing function.
 * @param {number} numOfCups Number of required cups of tea.
 * @return {Array<string>} Given amount of tea cups.
 **/

// tabs is an array of titles of each site open within the window

// we keep a record of the array inside the object

// When you join two windows into one window

// When you open a new tab at the end

// let&apos;s open a new tab for now

// When you close a tab

// get the tabs before the tab

// get the tabs after the tab

// join them together

// Let&apos;s create three browser windows

// Your mailbox, drive, and other work sites

// Social sites

//  Entertainment sites

// Now perform the tab opening, closing, and other operations

// Open a new tab for cat memes

// Close third tab in video window, and join

// only change code below this line

// only change code above this line

//ia.media-imdb.com/images/M/MV5BMjAxMzY3NjcxNF5BMl5BanBnXkFtZTcwNTI5OTM0Mw@@._V1_SX300.jpg&quot;,

//ia.media-imdb.com/images/M/MV5BMjIxNTU4MzY4MF5BMl5BanBnXkFtZTgwMzM4ODI3MjE@._V1_SX300.jpg&quot;,

//ia.media-imdb.com/images/M/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SX300.jpg&quot;,

//ia.media-imdb.com/images/M/MV5BNTM3OTc0MzM2OV5BMl5BanBnXkFtZTYwNzUwMTI3._V1_SX300.jpg&quot;,

//ia.media-imdb.com/images/M/MV5BMTYwOTEwNjAzMl5BMl5BanBnXkFtZTcwODc5MTUwMw@@._V1_SX300.jpg&quot;,

// :)

// Same, same; but different.

// Drop them elements.

// Is everyone being true?

// Complete the method below and implement the others similarly

// Remove all the values

// &quot;It&apos;s such a fine line between stupid, and clever.&quot;

// --David St. Hubbins

// I&apos;m a steamroller, baby

// What&apos;s in a name?

// LBH QVQ VG!

// Change the inputs below to test

// Here is your change, ma&apos;am.

// Example cash-in-drawer array:

// [[&quot;PENNY&quot;, 1.01],

// [&quot;NICKEL&quot;, 2.05],

// [&quot;DIME&quot;, 3.1],

// [&quot;QUARTER&quot;, 4.25],

// [&quot;ONE&quot;, 90],

// [&quot;FIVE&quot;, 55],

// [&quot;TEN&quot;, 20],

// [&quot;TWENTY&quot;, 60],

// [&quot;ONE HUNDRED&quot;, 100]]

// Good luck!

// Should print &quot;nom nom nom&quot;

// Should print &quot;Woof!&quot;

// Modify the code below this line

// =&gt; true

// Fix the code below so that it evaluates to true

/* jshint expr: true */

// After passing the challenge experiment with myString and see how the grouping works

// example crowd gathering

// Only change this line

//bit.ly/fcc-running-cats" class="img-responsive" alt="Three kittens running towards the camera.">

// const styles =

// write your code here

// initialize state here

// change code here

/* remove comment and change code below this line
    <h2>Welcome to React!</h2> <br>
    <p>Be sure to close all tags!</p>
    <hr>
    remove comment and change code above this line */

/* your code here */

/* change code here */

// &lt;&lt; change code here

// Redux:

// React:

// React-Redux:

// define the Container component here:

// complete the return statement:

// Change code below this line

// Change code above this line

// define ADD, addMessage(), messageReducer(), and store here:

// add handleChange() and submitMessage() methods here

/* render an input, button, and ul here */

// import React from &apos;react&apos;

// import ReactDOM from &apos;react-dom&apos;

// import { Provider, connect } from &apos;react-redux&apos;

// import { createStore, combineReducers, applyMiddleware } from &apos;redux&apos;

// import thunk from &apos;redux-thunk&apos;

// import rootReducer from &apos;./redux/reducers&apos;

// import App from &apos;./components/App&apos;

// const store = createStore(

//   rootReducer,

//   applyMiddleware(thunk)

// );

// ReactDOM.render(

//   <provider store="{store}">

//     <app>

//   </app></provider>,

//   document.getElementById(&apos;root&apos;)

// Redux Code:

// React Code:

// render the Provider here

// define the root reducer here

// don&apos;t mutate state here or the tests will fail

// Redux methods are available from a Redux object

// For example: Redux.createStore()

// Define the store here:

// Define an action here:

// Define an action creator here:

// Dispatch the action here:

// A list of strings representing tasks to do:

// an example todo argument would be &apos;Learn React&apos;,

// global count variable:

// dispatch request action here

// dispatch received data action here

// define a constant for increment action types

// define a constant for decrement action types

// define the counter reducer which will increment or decrement the state based on the action it receives

// define an action creator for incrementing

// define an action creator for decrementing

// define the Redux store here, passing in your reducers

// The main.scss file

/* Add your code below this line */

/* Add your code above this line */

// Create the scale here

// Call the scale with an argument here

// Padding between the SVG canvas boundary and the plot

// Create an x and y scale

// Returns 30

//jsonplaceholder.typicode.com/posts';

// test array:

// [1, 4, 2, 8, 345, 123, 43, 32, 5643, 63, 123, 43, 2, 55, 1, 234, 92]

// All inventory must be accounted for or you&apos;re fired!

// Example inventory lists

// Distance object returned

// Source: http://adripofjavascript.com/blog/drips/object-equality-in-javascript.html

// Create arrays of property names

// If number of properties is different,

// objects are not equivalent

// If values of same property are not equal,

// If we made it this far, objects

// are considered equivalent

// collection will hold our set

// this method will check for the presence of an element and return true or false

// this method will return all the values in the set

// write your add method here

// write your remove method here

// write your size method here

// case 1: target has no children, change code below this line

// find the target value and its parent

// count the children of the target to delete

// case 1: target has no children

// case 2: target has one child, change code below this line

// case 2: target has one child

// case 3: target has two children, change code below this line

// check if array is sorted

// generate a randomly filled array

// the var collection will hold the set

// this method will add an element to the set

// this method will remove an element from a set

// this method will return the size of the collection

// this method will return the union of two sets

// this method will return the intersection of two sets as a new set

// this method will return the difference of two sets as a new set

// this method will return the size of the set

// {1}

//Create a set with values 1, 2, 3, 4, &amp; 5

//Remove the value 2

//Remove the value 5

//Return the set

// Should return [ true, 3 ]

// only add code below this line

// You can do it!

// 3, 6

// empty

// 1

// 0,1,2,3,4,5,6

// 0

// noprotect

// Change after this line

// TODO add more tests

// good luck!
````

### LINKED ISSUE #37955: Parsing of translated challenges
Following on from the discussion in https://github.com/freeCodeCamp/freeCodeCamp/pull/37911 and conversations with @RandellDawson, I'm just writing it up for visibility.

Randell's idea is to simplify the process of keeping multiple translations in sync by modifying how challenges are parsed.  Instead of each language being independent, we could use the English markdown file as a template and add translated text from the other language's file.  That way the code (tests, seed and solutions) could be held in the English version and made available to all the other languages.

This would be achieved by modifying the `challenge-md-parser` for non-English languages.  It would first parse the English version, creating a challenge object, and then parse the other language and replace the challenge objects `description`, `instructions` (and the test `text` strings) with translated versions.

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
This treats translated challenges differently from English ones.  It takes the instructions, descriptions and test `text` from the translated challenge.  However, the seed, solution and `testString`s come from the English.  This means that bug fixes are automatically applied to all translations and translators can concentrate on language, without worrying about code.

Comments in the seed are replaced with translated versions if they exist.

Video challenges get the question and answers from the translated challenge.

Need to do
- [x] Create a real comment translation 'dictionary' @RandellDawson is working on this (done for Responsive Web Design)
- [x] ~~Make builds fail if any problems occur during parsing (missing comments only result in a warning, everything else should terminate the build)~~ the build no longer fails, so that the auditing can continue, but this should be reinstated before going to production
- [x] Remove the FOR TESTING commit
- [x] ~~`npm test:curriculum` should ignore locale, since all challenges will use the English tests (actually it had its own environment variable and the tests can be run using the results of `parseTranslation`)~~ I went in the opposite direction - tests are now controlled by `LOCALE` and can be run independently.  They should not fail if the English tests pass, but this should warn us if there are mismatches (challenges that do not appear in all languages, for example)
- [x] certificates need handling separately
- [ ] ~~check if `.language.md` extensions are necessary, if not https://github.com/freeCodeCamp/freeCodeCamp/pull/39024 can be closed.~~ To avoid changing too much in one go, I'm sticking with them for now.

Would be nice
- [x] Warn if a comment is not in the dictionary

Closes #37955
Closes #37599

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
This is mostly a write-up of what @ahmadabdolsaheb and I discussed.  

Having translated challenges with descriptions that don't match the tests is something we'd need to avoid.  This could happen if the English challenge gets updated with a new description and tests (and no-one updates the translations).  As soon as the changes to the English are live, the translated challenges would start using the new tests and there would be a mismatch.

One approach would be to automatically create translation issues whenever an English challenge's text changes (via Github Actions, maybe) and only allow the changes into production once all the translations are in.

A more elaborate (and difficult to implement) approach would be to version the challenges and tie the translations to a particular version. That way a new English version could be quickly pushed to live, with the translations using the tests from the old version until a new translation PR is merged.

**Comment by ojeytonwilliams** [comment]
To expand a bit, I think this PR would cause most problems when the `text` *and* `testString` are both updated in the English.  Reason being, this is likely a change in the requirements to complete the challenge, not simply a bug fix.

Any English language PR which changes neither the `testString`s nor the seed has no affect the translated challenges at all, so is not breaking.  An issue should be created, since it's a hint that the translations may need changing, but the English PR can be merged regardless.

If the description and the seed are changed, it could mean that the translated description now refers to something that has been removed from the seed.  This would be breaking.

In summary:

Seed change or `testString` change combined with any other text changes could be breaking. Everything else should not be.

**Comment by ojeytonwilliams** [comment]
@raisedadead I've cleaned it up, given it another look and it seems fine.  It probably should target the `next-chinese` branch, right?  If so, is there any reason to keep a PR targeting master?

**Comment by raisedadead** [comment]
> It probably should target the `next-chinese` branch, right? If so, is there any reason to keep a PR targeting master?

Well if you are able to rebase this on top of `next-chinese`, then sure, if not you can leave it as is.

**Comment by ojeytonwilliams** [comment]
@raisedadead I had to remove 10 Chinese challenges that were no longer in the English, but having done that, the Chinese curriculum only has 1467 challenges (to the 1533 English ones).

Also, there are a few challenges where the tests don't match.  For now I've suppressed the warnings so the site can be built.

**Comment by RandellDawson** [comment]
> the Chinese curriculum only has 1467 challenges (to the 1533 English ones).

Do you have a list of the English challenges that need a machine translation, so the Chinese team can review and make suggestions for correction?  I can do the translations for those challenges in a separate PR or we can just create an issue listing all the challenges that need to be converted to Chinese and have the Chinese team or other contributors to "real" translations of them.

**Comment by ojeytonwilliams** [comment]
> Do you have a list of the English challenges that need a machine translation

Not yet.  I'll make a list of all the challenges with mismatched tests and all the ones that don't exist in the Chinese.

**Comment by raisedadead** [comment]
Okay, I understand that the Chinese curriculum has obviously divulged. The idea is to be able to make the translations live one cert at a time. We know that all the Web Design Cert stuff should be in order. And currently the @freeCodeCamp/chinese team is working on the Basic JavaScript sections.

So, we could modify the parser to be able to pick only certs that we choose?

**Comment by raisedadead** [comment]
After landing this PR, we would like to be able to land one certificate at a time on the .dev so, that it unblocks the Chinese mods team to start working on the next one in tandem.

**Comment by ojeytonwilliams** [comment]
> So, we could modify the parser to be able to pick only certs that we choose?

Absolutely.  So, should it just default to the English unless it's one of the allowed certs?

**Comment by ojeytonwilliams** [comment]
The Chinese challenges that are not in the English curriculum, with explanations (most have been renamed):

```
02-javascript-algorithms-and-data-structures/es6/understand-the-differences-between-import-and-require.chinese.md
renamed to
reuse-javascript-code-using-import

02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-with-the-rest-operator-to-reassign-array-elements.chinese.md
renamed to
use-destructuring-assignment-with-the-rest-parameter-to-reassign-array-elements

02-javascript-algorithms-and-data-structures/es6/use-export-to-reuse-a-code-block.chinese.md
renamed to 
use-export-to-share-a-code-block

02-javascript-algorithms-and-data-structures/es6/use-the-rest-operator-with-function-parameters.chinese.md
renamed to
use-the-rest-parameter-with-function-parameters

02-javascript-algorithms-and-data-structures/es6/write-concise-object-literal-declarations-using-simple-fields.chinese.md
renamed to
write-concise-object-literal-declarations-using-object-property-shorthand

02-javascript-algorithms-and-data-structures/es6/write-higher-order-arrow-functions.chinese.md
moved to
02-javascript-algorithms-and-data-structures/functional-programming/use-higher-order-functions-map-filter-or-reduce-to-solve-a-complex-problem

03-front-end-libraries/react/manage-updates-with-lifecycle-methods.chinese.md
removed from the curriculum

08-coding-interview-prep/data-structures/remove-from-a-set.chinese.md
removed, merged into
create-a-set-class
```

All the challenges where the number of tests in the English doesn't match the number in the Chinese:
> Apply a Style Until a Condition is Met with @while
Create a Media Query
Add Comments in JSX
Request Header Parser Microservice
Make an Image Responsive
Find Characters with Lazy Matching
Create Strings using Template Literals
Add Placeholder Text to a Text Field
Manage State Locally First
Create a Bulleted Unordered List
Inherit Behaviors from a Supertype
Check if an Object has a Property
Understanding the Differences between the freeCodeCamp and Browser Console
Mutations
Repeat a String Repeat a String
Create a Set Class
Match Characters that Occur Zero or More Times
Understand Own Properties
Use Destructuring Assignment to Assign Variables from Nested Objects
Iterate Through the Keys of an Object with a for...in Statement
Use Destructuring Assignment to Assign Variables from Objects
Use Destructuring Assignment to Pass an Object as a Function's Parameters
Give Sibling Elements a Unique Key Attribute
Steamroller
Use getters and setters to Control Access to an Object
Remove Items Using splice()
Use Closure to Protect Properties Within an Object from Being Modified Externally
Use the reduce Method to Analyze Data
Pass Props to a Stateless Functional Component
Positive and Negative Lookahead
Restrict Possible Usernames
Use Capture Groups to Search and Replace
Remove Elements from a Linked List by Index
Remove Elements from a Linked List
Use Clockwise Notation to Specify the Margin of an Element
Use Array.map() to Dynamically Render Elements
Use CSS Selectors to Style Elements
Escape Sequences in Strings
Use State to Toggle an Element
Find the Length of a String
Use the u Tag to Underline Text
Record Collection
Word Blanks
Problem 2: Even Fibonacci Numbers

Finally, all the challenges that are missing from the Chinese curriculum, this includes some which have been renamed:

>  '02-javascript-algorithms-and-data-structures/regular-expressions/check-for-mixed-grouping-of-characters',
  '02-javascript-algorithms-and-data-structures/es6/complete-a-promise-with-resolve-and-reject',
  '08-coding-interview-prep/data-structures/check-if-binary-search-tree',
  '02-javascript-algorithms-and-data-structures/es6/create-a-javascript-promise',
  '04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method',
  '02-javascript-algorithms-and-data-structures/es6/create-a-module-script',
  '02-javascript-algorithms-and-data-structures/es6/handle-a-fulfilled-promise-with-then',
  '02-javascript-algorithms-and-data-structures/es6/handle-a-rejected-promise-with-catch',
  '02-javascript-algorithms-and-data-structures/es6/reuse-javascript-code-using-import',
  '02-javascript-algorithms-and-data-structures/functional-programming/use-higher-order-functions-map-filter-or-reduce-to-solve-a-complex-problem',
  '08-coding-interview-prep/rosetta-code/cumulative-standard-deviation',
  '08-coding-interview-prep/rosetta-code/cusip',
  '02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-extract-values-from-objects',
  '08-coding-interview-prep/rosetta-code/cut-a-rectangle',
  '02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-with-the-rest-parameter-to-reassign-array-elements',
  '02-javascript-algorithms-and-data-structures/es6/use-export-to-share-a-code-block',
  '02-javascript-algorithms-and-data-structures/es6/use-the-rest-parameter-with-function-parameters',
  '02-javascript-algorithms-and-data-structures/es6/write-concise-object-literal-declarations-using-object-property-shorthand',
  '08-coding-interview-prep/rosetta-code/dot-product',
  '08-coding-interview-prep/rosetta-code/generator-exponential',
  '08-coding-interview-prep/rosetta-code/k-d-tree',
  '08-coding-interview-prep/rosetta-code/kaprekar-numbers',
  '02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion',
  '08-coding-interview-prep/rosetta-code/knights-tour',
  '08-coding-interview-prep/rosetta-code/largest-int-from-concatenated-ints',
  '08-coding-interview-prep/rosetta-code/last-friday-of-each-month',
  '08-coding-interview-prep/rosetta-code/leap-year',
  '08-coding-interview-prep/rosetta-code/least-common-multiple',
  '08-coding-interview-prep/rosetta-code/left-factorials',
  '08-coding-interview-prep/rosetta-code/sort-an-array-of-composite-structures',
  '08-coding-interview-prep/rosetta-code/sort-disjoint-sublist',
  '08-coding-interview-prep/rosetta-code/sort-stability',
  '08-coding-interview-prep/rosetta-code/sort-using-a-custom-comparator',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-bead-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-bogosort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-cocktail-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-comb-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-gnome-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-pancake-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-permutation-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-shell-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-stooge-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-strand-sort',
  '08-coding-interview-prep/rosetta-code/soundex',
  '08-coding-interview-prep/rosetta-code/spiral-matrix',
  '02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers',
  '08-coding-interview-prep/rosetta-code/split-a-character-string-based-on-change-of-character',
  '08-coding-interview-prep/rosetta-code/state-name-puzzle',
  '08-coding-interview-prep/rosetta-code/stern-brocot-sequence',
  '08-coding-interview-prep/rosetta-code/straddling-checkerboard',
  '08-coding-interview-prep/rosetta-code/stream-merge',
  '08-coding-interview-prep/rosetta-code/strip-control-codes-and-extended-characters-from-a-string',
  '08-coding-interview-prep/rosetta-code/subleq',
  '08-coding-interview-prep/rosetta-code/sudoku',
  '08-coding-interview-prep/rosetta-code/sum-digits-of-an-integer',
  '08-coding-interview-prep/rosetta-code/sum-multiples-of-3-and-5',
  '08-coding-interview-prep/rosetta-code/sum-of-a-series',
  '08-coding-interview-prep/rosetta-code/sum-of-squares',
  '08-coding-interview-prep/rosetta-code/sum-to-100',
  '08-coding-interview-prep/rosetta-code/sutherland-hodgman-polygon-clipping',
  '08-coding-interview-prep/rosetta-code/symmetric-difference'

Apologies for the slight differences in naming (titles, paths, etc), I can be more consistent if needed, but they should all be searchable!

**Comment by ojeytonwilliams** [comment]
I've noticed some challenge files are not in the form `/path/to/name.lang.md` e.g. 
```
curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/stand-in-line.english.md
```
is fine, but
```
curriculum/challenges/english/08-coding-interview-prep/rosetta-code/knapsack-problem-0-1.md
```
is not.

I'm updating my PR to throw errors when it encounters these and I'll create PRs fixing them, but some of the above lists may have false positives.  I'll update the lists ASAP.

**Comment by RandellDawson** [comment]
All files should have the language extension + `md`.

**Comment by RandellDawson** [comment]
@ankush0512  Just wanted to let you know, that I have already have PRs that consolidate the comments (they are currently being reviewed) and will be merged shortly.  I have also created the applicable translation lookup object based on the consolidated comments that were left.  I will also be taking care of the other curriculum sections as we will be  working through one section at a time.

**Comment by raisedadead** [comment]
I bet we are getting along real close with this? 

I have landed all the Chinese i18n progress so far (for the first cert). I have also rebased this PR off of master and changed the target.

**Comment by ojeytonwilliams** [comment]
> I bet we are getting along real close with this?

Yep, pretty close.  I need to review Randell's PRs though, so I'll get on that.

**Comment by ojeytonwilliams** [comment]
@raisedadead btw, do you want the un-validated certs (i.e. non Responsive Web Design) to disappear from the curriculum entirely or just default to English?

**Comment by raisedadead** [comment]
I would be fine with them disappearing.

**Comment by ojeytonwilliams** [comment]
> I would be fine with them disappearing.

I'm looking into this, but it has to be done carefully or Gatsby breaks.  Might take a little while.

**Comment by raisedadead** [comment]
> I'm looking into this, but it has to be done carefully or Gatsby breaks. Might take a little while.

OK, I would defer to your best judgement. If it becomes too annoying to implement, let us default to English.

Let me know if I can be of any help.

**Comment by ojeytonwilliams** [comment]
@raisedadead> Let me know if I can be of any help.

I think it's more or less okay. The last two commits do *most* of it, but there's still the issue that all the certs appear in the settings page, even though they're gone from the curriculum.  Finally I still need to handle the certificate challenges, as they have a different structure for the tests, but it should be easy to do.

If you could take a look and let me know what needs doing differently, that would be a huge help.

File: client/plugins/fcc-source-challenges/create-challenge-nodes.js
Comment (RandellDawson) [inline client/plugins/fcc-source-challenges/create-challenge-nodes.js]: What about creating an object (i.e. `emptyFiles`) and stick it after the `?`.  Then before the return statement, you could use something like the following to create the object.
```
const emptyFiles = ['html', 'js', 'jsx'].reduce((obj, type) => {
  const properties = ['key', 'ext', 'name', 'contents', 'head', 'tail'];
  const keyValuesObj = properties.reduce((obj, prop) => ({ ...obj, [prop]: ''}), {});
  return { ...obj, [`index${type}`]: keyValuesObj };
}, {});
```

Comment (ojeytonwilliams) [reply client/plugins/fcc-source-challenges/create-challenge-nodes.js]: Good point, it is rather ugly.  I'll sort that out tomorrow!

Comment (ojeytonwilliams) [reply client/plugins/fcc-source-challenges/create-challenge-nodes.js]: I based my solution on yours, but decided against the nested reduce for reasons of readability.  Is this version okay?

File: client/plugins/fcc-source-challenges/create-challenge-nodes.js
Comment (RandellDawson) [inline client/plugins/fcc-source-challenges/create-challenge-nodes.js]:
```suggestion
const EMPTY_FILES = ['indexhtml', 'indexjs', 'indexjsx'].reduce(
  (acc, curr) => ({ ...acc, [curr]: FILE_CONTENTS })
, {});
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
How about a slightly more concise version?

File: client/plugins/fcc-source-challenges/create-challenge-nodes.js
Comment (ojeytonwilliams) [reply client/plugins/fcc-source-challenges/create-challenge-nodes.js]: I like it.  Could you put this comment on the latest commit?  Github's telling me I can't commit the suggestion because it's 'outdated'.

**Comment by ojeytonwilliams** [comment]
@raisedadead the basic functionality is all there, for the first cert at least.  I went ahead and removed the untranslated/unaudited certs from the settings, in the last commit.

**Comment by RandellDawson** [comment]
@ojeytonwilliams So is the plan to merge this one and make sure nothing "breaks" on the dev site and then remove all but the Description, Instructions, and `text`s of the Tests section in a separate PR and make sure nothing breaks?

**Comment by ojeytonwilliams** [comment]
The way I see it going is this:

1. I'll clean it up (there are way too many commits at the moment)
2. Test it locally
3. Test on .dev (or a Chinese specific domain?)

Removing the unnecessary parts from the Chinese challenges can happen whenever we want. The parser can already cope, the latest test just confirms that. Presumably we'll remove them after we've used it in production for a while and haven't run into any insurmountable problems.

**Comment by raisedadead** [comment]
> Removing the unnecessary parts from the Chinese challenges can happen whenever we want.

We have a snapshot already (`backup-*` branches). However, I would still want to do it in a separate PR.

**Comment by raisedadead** [comment]
Also, I am ready to pull this down and deploy whenever you want me to. I will prepare the Vm's etc, when you give me the go ahead.

**Comment by ojeytonwilliams** [comment]
@raisedadead it should be good to go. There was a bug, now squashed, where the English settings only showed Responsive Web Design, so it should be QAed in both English and Chinese.  Just to be sure.

File: client/plugins/fcc-source-challenges/create-challenge-nodes.js
Comment (ojeytonwilliams) [reply client/plugins/fcc-source-challenges/create-challenge-nodes.js]: When I cleaned things up, I included this with one last tweak: it now clones `FILE_CONTENTS`.

**Comment by raisedadead** [comment]
I have started the work on deploying this at `chinese.freecodecamp.dev` via https://github.com/freeCodeCamp/nginx-config/pull/1, I will have something by morning.

**Comment by ojeytonwilliams** [comment]
I avoided using a parser in this because I was worried that it would reformat the code, but 
 @RandellDawson's been looking at using Babel and a css parser to remove comments from user code.  If that works and has no unwanted side-effects, we can consider using that instead of the regex.

**Comment by moT01** [comment]
I have blocked this for now. See the [pinned code freeze issue](https://github.com/freeCodeCamp/freeCodeCamp/issues/38308)

**Pull request merged**
